### PR TITLE
Issue62

### PR DESCRIPTION
**Pull request by jpinedaf**
Trying to add capability for gridding individual sessions or session range (Issue #62). 
@rfriesen @low-sky 
What do you think of this? Does this make sense?

### CODE REVIEW

**Comment by rfriesen** [comment]
Using session ranges is definitely an easy way to adjust the final map size. Would it be possible to include a list of sessions, in cases where maps are made or added to out of order?

**Comment by jpinedaf** [comment]
are you thinking about something like this? If map Ceph_11 is observed in sessions: 12, 15, and 18, but other maps of Ceph have been observed in session 16. If that is the case, what kind of extension would you put to the FITS file?

File: GAS/gridregion.py [hunk 1]
```diff
@@ -114,9 +114,20 @@ def griddata(pixPerBeam = 3.0,
...
+        filelist = glob.glob(rootdir+'/'+region+'/'+dirname+'/*fits')
+        file_extension='_all'
+    else:
+        filelist = []
+        for scan_i in range(startSession, endSession+1):
+            filelist.extend(glob.glob(rootdir+'/'+region+'/'+dirname+'/*_sess'+str(scan_i)+'.fits'))
+        if startSession == endSession:
+            file_extension='_sess'+str(startSession)
+        else:
+            file_extension='_sess'+str(startSession)+'-sess'+str(endSession)
```
Comment (low-sky) [inline GAS/gridregion.py]: `sess%(start)s-sess%(ends)s' % {'start':Session[0],'ends':Session[-1]}`

File: GAS/gridregion.py [hunk 2]
```diff
@@ -114,9 +114,20 @@ def griddata(pixPerBeam = 3.0,
...
              doBaseline = True,
-             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)]):
+             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)],
+             startSession = 1, endSession = 100):
+    if (startSession == 1) and (endSession == 100):
+        filelist = glob.glob(rootdir+'/'+region+'/'+dirname+'/*fits')
+        file_extension='_all'
+    else:
+        filelist = []
+        for scan_i in range(startSession, endSession+1):
```
Comment (low-sky) [inline GAS/gridregion.py]: This would become
 `for scan_i in Session:`

File: GAS/gridregion.py [hunk 3]
```diff
@@ -114,9 +114,20 @@ def griddata(pixPerBeam = 3.0,
              dirname = 'NGC1333_NH3_11',
              startChannel = 1024, endChannel = 3072,
              doBaseline = True,
-             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)]):
+             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)],
+             startSession = 1, endSession = 100):
+    if (startSession == 1) and (endSession == 100):
```
Comment (low-sky) [inline GAS/gridregion.py]: `if not Session:`

File: GAS/gridregion.py [hunk 4]
```diff
@@ -114,9 +114,20 @@ def griddata(pixPerBeam = 3.0,
              dirname = 'NGC1333_NH3_11',
              startChannel = 1024, endChannel = 3072,
              doBaseline = True,
-             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)]):
+             baselineRegion = [slice(512,1024,1),slice(3072,3584,1)],
+             startSession = 1, endSession = 100):
```
Comment (low-sky) [inline GAS/gridregion.py]: `Sessions = None`

**Comment by low-sky** [comment]
I think @rfriesen's idea here is spot on.  This can be a list of sessions rather than (or including) a range.  It does allow fine tuning of the sessions included in the map.  I think we should have this be a list or array, for example `Sessions = [12, 15, 18]` or `Sessions = range(12, 19)` and then iterate over this list.  

The remaining questions would be defaults.  I think we want `Sessions = None` as the default and then the file_extension would maybe be start and end?  I've added some some code comments in the PR.

`

File: GAS/gridregion.py (hunk 1 above)
Comment (low-sky) [inline GAS/gridregion.py]: Or, more simply,
`'sess{0}-sess{1}'.format(Session[0],Session[-1])`

**Comment by jpinedaf** [comment]
That is great I'm implementing the changes, however I wonder: how would you discriminate between 
`Sessions=[1,5,9]` and `Sessions=range(1,10)` if we only keep the start-end session numbers? On the other hand, including **all** session numbers in the extension sounds terrible.

**Comment by low-sky** [comment]
Yeah, I saw that as an important case, but it would get unwieldy.  I came up with no solutions past the awful (all the session numbers) and the opaque (unique reduction numbers) I erred on the side of brevity and understanding the side effects.  Another enhancement would be to add `HISTORY` cards to the FITS header.

**Comment by jpinedaf** [comment]
Added the Session numbers used in the imaging. 

@low-sky @keflavich Do you know how could I get the version of the pipeline that we are using? Usually I would say `GAS.__version__` but how does it work from within the package?

**Comment by keflavich** [comment]
@jpinedaf if it's installed or `develop`'d `GAS.__version__` should work - does it not?  That might be some sort of issue with astropy-helpers

**Comment by jpinedaf** [comment]
but if you do `import GAS as G`, would that break the `GAS.__version__`?

**Comment by keflavich** [comment]
no, it shouldn't, afaik.  You'd just use `G.__version__` instead

**Comment by jpinedaf** [comment]
I'm trying to use the following command within the gridding function to store the pipeline version in the FITS header
```
hdr.add_history('Using GAS pipeline version {0}'.format(GAS.__version__))
```

**Comment by jpinedaf** [comment]
This is the original line of code https://github.com/GBTAmmoniaSurvey/GAS/commit/7996bb4b560079f5399689bff123ed13ea271f07#diff-6f9fa0895a507f41d8a38636de4e6ddeR259
I'm trying to make sure it works

**Comment by keflavich** [comment]
I see the issue now... maybe it's better to do `from . import __version__`.  I think that should work; that style of import is used within astropy

**Pull request merged**